Integrate Rewindable outputs into Transactions

philipr-za opened the pull request:
## Description

This PR adds the ability for outputs generated in transactions to be generated using rewind keys that can allow for a partial or full rewinding or a range proof. The Output Manager Service now derives two extra keys using the master_seed key and two special branches in the key manager to be used as a `rewind_key` and `rewind_blinding_key`. All transactions that are produced by the wallet (Coinbases, Sender and Recipient transactions) will now be rewindable using these keys. The public versions allow for the first level of rewinding that reveals the proof message and value and the private versions allow for a full rewind to also reveal the spending key.

It is crucial to note that the private `rewind_key` and `rewind_blinding_key` are both extremely sensitive. Because we use the same keys for all outputs generated by a given wallet if these keys are compromised the entire wallet is compromised, not just the spending key for a single UTXO. Thus a focus of the updates to the Transaction negotiation protocols and CoinbaseBuilder in the `tari_core` crate was to ensure that the private rewind keys are only ever used in memory and are never serialised in the output of the protocols. They are only ever included in the serialised output as fully formed range proofs which are secure. 

A Public API call was added to the Output Manager Service to return the public rewind keys for use as view keys and to use in testing.

The PR does not provide wallet restoration or the functionality to use the view keys as yet.

Summary:
- Added `as_rewindable_transaction_output` to Unblinded Output
- Added two levels of rewinding to `TransactionOutput`
- Added being able to produce a rewindable output to `ReceiverTransactionProtocol`
- Added being able to produce a rewindable change output to `SenderTransactionInitializer` and the `SingleReceiverTransactionProtocol`
- Added being able to produce rewindable Coinbase outputs to the `CoinbaseBuilder`
- Updated OutputManagerService to produce the two rewind keys from its embedded KeyManager
- Updated requesting the Recipients half of a inbound transaction to use rewindable outputs
- Updated producing the Sender half of an outbound transaction to use rewindable outputs and change.
- Updated producing a Coinbase transaction to create a rewindable output
- Add API to get public rewind keys from Output Manager to allow for testing rewind and providing view keys.

## Motivation and Context
This is the first half of what is needed for wallet restoration, this PR will mean that outputs on the blockchain will now be rewindable with keys generated from the wallet's master seed.

## How Has This Been Tested?
Tests were provided for all the elements added to 'tari_core' and tests were added and updated in the wallet to check for rewindability

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

delta1 commented:
Really great work! 👏  I'm digging into it, but in the meanwhile there's a small clippy/build issue for an unused variable

delta1 commented on base_layer/core/src/transactions/coinbase_builder.rs:
```diff
@@ -266,6 +282,39 @@ mod test {
+        let p = TestParams::new();
+        let (builder, rules, factories) = get_builder();
+        let builder = builder
+            .with_block_height(42)
```
good ol' answer to everything 🤣

delta1 reviewed (dismissed):
LGTM

philipr-za commented:
Rebase and deal with a conflict

stringhandler commented on base_layer/core/src/transactions/transaction.rs:
```diff
@@ -223,6 +231,39 @@ impl UnblindedOutput {
+            proof,
+        };
+        // A range proof can be constructed for an invalid value so we should confirm that the proof can be verified.
+        if !output.verify_range_proof(&factories.range_proof)? {
```
Might be worth skipping this in Release mode. Is it not cheaper to verify `self.value` directly or is there another way of creating an invalid rangeproof?

philipr-za replied:
🤷‍♂️  This is from the original implementation of UnblindedOutput. If we are confident that the value in self.value cannot be invalid I guess we can remove this here and in 'as_transaction_output`

stringhandler commented on base_layer/core/src/transactions/transaction.rs:
```diff
@@ -353,7 +394,38 @@ impl TransactionOutput {
+    pub fn rewind_range_proof(
+        &self,
+        prover: &RangeProofService,
+        public_rewind_key: &PublicKey,
```
nit: Perhaps `rewind_public_key` and `rewind_blinding_public_key` is a better name? We don't want to suggest that these values are published somewhere

stringhandler commented on base_layer/core/src/transactions/transaction.rs:
```diff
@@ -353,7 +394,38 @@ impl TransactionOutput {
+    }
+
+    /// Attempt to rewind the range proof to reveal the proof message and committed value
+    pub fn rewind_range_proof(
```
```suggestion
    pub fn rewind_range_proof_value_only(
```
Not sure what a better name would be, but my intention here is to verbalise how this method differs from `full_rewind_range_proof`

stringhandler commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -74,6 +87,10 @@ use tokio::{
 const LOG_TARGET_STRESS: &str = "stress_test::output_manager_service";
 
+const KEY_MANAGER_COINBASE_BRANCH_KEY: &str = "coinbase";
+const KEY_MANAGER_REWIND_BRANCH_KEY: &str = "rewind";
```
```suggestion
const KEY_MANAGER_RECOVERY_VIEWONLY_BRANCH_KEY: &str = "recovery_viewonly";
```

stringhandler commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -74,6 +87,10 @@ use tokio::{
 
+const KEY_MANAGER_COINBASE_BRANCH_KEY: &str = "coinbase";
+const KEY_MANAGER_REWIND_BRANCH_KEY: &str = "rewind";
+const KEY_MANAGER_REWIND_BLINDING_BRANCH_KEY: &str = "rewind_blinding";
```
```suggestion
const KEY_MANAGER_RECOVERY_BLINDING_BRANCH_KEY: &str = "recovery_blinding";
```

stringhandler reviewed (commented):
Happy with the approach. I think there are some places where the naming could be changed for clarity, specifically using `recovery` instead of `rewind`. This is a subjective view and I don't want to go down the bike shedding path , so I'm not too fussed about it

philipr-za commented:
> Happy with the approach. I think there are some places where the naming could be changed for clarity, specifically using `recovery` instead of `rewind`. This is a subjective view and I don't want to go down the bike shedding path , so I'm not too fussed about it

My issue with that is that we are using the rewinding for recovery but we can also use the rewinding for other purposes when we put stuff into the proof message so these methods might not only be used for recovery purposes.

stringhandler commented:
Fair enough

philipr-za committed:
Integrate Rewindable outputs into Transactions

This PR adds the ability for outputs generated in transactions to be generated using rewind keys that can allow for a partial or full rewinding. The Output Manager Service now derives two extra keys using the master_seed key and two special branches in the key manager to be used as a `rewind_key` and `rewind_blinding_key`. All transactions that are produced by the wallet (Coinbases, Sender and Recipient transactions) will now be rewindable using these keys. The public versions allow for the first level of rewinding that reveals the proof message and value and the private versions allow for a full rewind to also reveal the spending key.

It is crucial to note that the private `rewind_key` and `rewind_blinding_key` are both extremely sensitive. Because we use the same keys for all outputs generated by a given wallet if these keys are compromised the entire wallet is compromised, not just the spending key for a single UTXO. Thus a focus of the updates to the Transaction negotiation protocols and CoinbaseBuilder in the `tari_core` crate was to ensure that the private rewind keys are only ever used in memory and are never serialised in the output of the protocols. They are only ever included in the serialised output as fully formed range proofs which are secure. 

A Public API call was added to the Output Manager Service to return the public rewind keys for use as view keys and to use in testing.

The PR does not provide wallet restoration or the functionality to use the view keys as yet.
